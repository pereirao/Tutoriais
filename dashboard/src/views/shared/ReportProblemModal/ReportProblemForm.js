// @flow

import React from 'react';
import {connect} from 'react-redux';
import _ from 'lodash';
import I18n from '../../../localization';
import {comment_actions, comment_helpers, comment_selectors} from '../../../business/comment';
import type {DeliveryMethodType, Order} from '../../../business/order';
import {order_actions, order_selectors} from '../../../business/order';
import style_sheet from './ReportProblemModal.module.css';
import {MBForm, MBFormSubmitButton} from '../../elements/MBForm';
import {FreeResponseField, OrderItemField} from '../form_fields';

const {exceptionComment} = comment_helpers;

const problem_forms = {
  out_of_stock: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel key="which_label">{I18n.t('ui.modal.report_problem.label.which_item')}</FieldLabel>
      <OrderItemField key="order_item_field" order={order}/>
      <FieldLabel key="replacement_item_label">{I18n.t('ui.modal.report_problem.label.replacement_item')}</FieldLabel>
      <FreeResponseField key="replacement_item" name="replacement_item"/>
      <FieldLabel key="replacement_price_label">{I18n.t('ui.modal.report_problem.label.replacement_price')}</FieldLabel>
      <FreeResponseField key="replacement_price" name="replacement_price"/>
    </ReportProblemFormWrapper>
  ),
  incorrect_pricing: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel key="which_label">{I18n.t('ui.modal.report_problem.label.which_item')}</FieldLabel>
      <OrderItemField key="order_item_field" order={order}/>
      <FieldLabel key="correct_price_label">{I18n.t('ui.modal.report_problem.label.incorrect_pricing')}</FieldLabel>
      <FreeResponseField key="correct_price" name="correct_price"/>
    </ReportProblemFormWrapper>
  ),
  incorrect_listing: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel key="which_label">{I18n.t('ui.modal.report_problem.label.which_item')}</FieldLabel>
      <OrderItemField key="order_item_field" order={order}/>
      <FieldLabel key="incorrect_listing_label">{I18n.t('ui.modal.report_problem.label.incorrect_listing')}</FieldLabel>
      <FreeResponseField key="incorrect_listing" name="incorrect_listing"/>
    </ReportProblemFormWrapper>
  ),
  incomplete_address: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel
        key="incomplete_address_label">{I18n.t('ui.modal.report_problem.label.incomplete_address')}</FieldLabel>
      <FreeResponseField key="incomplete_address" name="incomplete_address"/>
    </ReportProblemFormWrapper>
  ),
  suspected_fraud: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel key="suspected_fraud_label">{I18n.t('ui.modal.report_problem.label.suspected_fraud')}</FieldLabel>
      <FreeResponseField name="suspected_fraud"/>
    </ReportProblemFormWrapper>
  ),
  driver_unavailable: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel
        key="driver_unavailable_why_label">{I18n.t('ui.modal.report_problem.label.driver_unavailable_why')}</FieldLabel>
      <FreeResponseField name="driver_unavailable_why"/>
      {
        //TODO: DatePicker "When will your driver be available"
        //Comment Field -- If next day should we close you until then?
        //Message "we will contact the customer and let them know that it will either be rescheduled or canceled"
      }
    </ReportProblemFormWrapper>
  ),
  failed_delivery: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel key="failed_delivery_label">{I18n.t('ui.modal.report_problem.label.failed_delivery')}</FieldLabel>
      <FreeResponseField name="failed_delivery"/>
      {
        //TODO: Dropdown Failed Delivery reasons ("Customer wasn't present" "Doorman would not accept the order" "Address could not be found" "Customer refused delivery")
      }
    </ReportProblemFormWrapper>
  ),
  other: (problem_type, order, isMinibarClosed) => (
    <ReportProblemFormWrapper
      order={order}
      problem_type={problem_type}
      isMinibarClosed={isMinibarClosed}>
      <FieldLabel key="other_label">{I18n.t('ui.modal.report_problem.label.other')}</FieldLabel>
      <FreeResponseField name="other"/>
    </ReportProblemFormWrapper>
  )
};

const FieldLabel = ({children}) => <p className={style_sheet.fieldLabel}>{children}</p>;

const ReportProblemFormWrapperSTP = state => ({
  is_order_updating: order_selectors.isUpdating(state),
  is_comment_adding: comment_selectors.isFetching(state)
});
const ReportProblemFormWrapperDTP = {
  updateOrder: order_actions.updateOrder,
  addComment: comment_actions.addComment
};
const connectReportProblemFormWrapper = connect(ReportProblemFormWrapperSTP, ReportProblemFormWrapperDTP);
type ReportProblemFormWrapperProps = {
  problem_type: DeliveryMethodType,
  order: Order,
  isMinibarClosed: () => void,
  updateOrder: (string, Object) => void,
  addComment: (string, string) => void,
  is_order_updating: boolean,
  is_comment_adding: boolean,
  children: any
};
const ReportProblemFormWrapper = connectReportProblemFormWrapper(({
                                                                    problem_type,
                                                                    order,
                                                                    isMinibarClosed,
                                                                    updateOrder,
                                                                    addComment,
                                                                    is_order_updating,
                                                                    is_comment_adding,
                                                                    children
                                                                  }: ReportProblemFormWrapperProps) => {
  const handleSubmit = values => {
    const exception = {type: problem_type, description: problem_type, metadata: values};
    updateOrder(order.id, {state: 'exception', exception});
    addComment(order.id, exceptionComment(exception));
    isMinibarClosed();
  };
  const validateForm = fields => _.every(fields, 'value');
  return (
    <MBForm
      key={problem_type}
      onSubmit={handleSubmit}
      validateSubmittable={validateForm}
      className={style_sheet.reportProblemForm}>
      {children}
      <p className={style_sheet.formNote}>{I18n.t('ui.modal.report_problem.notify_minibar')}</p>
      <MBFormSubmitButton
        is_loading={is_order_updating || is_comment_adding}
        key="submit"
        text={I18n.t('ui.modal.report_problem.submit')}/>
    </MBForm>
  );
});

type ReportProblemFormProps = {
  problem_type: DeliveryMethodType,
  order: Order,
  isMinibarClosed: () => void
};
const ReportProblemForm = ({
                             problem_type,
                             order,
                             isMinibarClosed
                           }: ReportProblemFormProps) => {
  if (!problem_type) return null;
  return problem_forms[problem_type](problem_type, order, isMinibarClosed);
};

export default ReportProblemForm;
