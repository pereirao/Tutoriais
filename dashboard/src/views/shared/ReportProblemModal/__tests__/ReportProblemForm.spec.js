import React from 'react';

import ReportProblemForm from '../ReportProblemForm';

it('renders form for out_of_stock problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="out_of_stock" />)).toMatchSnapshot();
});
it('renders form for incorrect_pricing problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="incorrect_pricing" />)).toMatchSnapshot();
});
it('renders form for incorrect_listing problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="incorrect_listing" />)).toMatchSnapshot();
});
it('renders form for incomplete_address problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="incomplete_address" />)).toMatchSnapshot();
});
it('renders form for suspected_fraud problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="suspected_fraud" />)).toMatchSnapshot();
});
it('renders form for driver_unavailable problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="driver_unavailable" />)).toMatchSnapshot();
});
it('renders form for failed_delivery problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="failed_delivery" />)).toMatchSnapshot();
});
it('renders form for other problem_type', () => {
  expect(shallow(<ReportProblemForm problem_type="other" />)).toMatchSnapshot();
});
