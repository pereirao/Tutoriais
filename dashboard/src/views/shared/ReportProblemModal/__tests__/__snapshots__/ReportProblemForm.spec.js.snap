// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form for driver_unavailable problem_type 1`] = `
<Connect(Component)
  problem_type="driver_unavailable"
>
  <FieldLabel>
    Why is your driver unavailable?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="driver_unavailable_why"
  />
</Connect(Component)>
`;

exports[`renders form for failed_delivery problem_type 1`] = `
<Connect(Component)
  problem_type="failed_delivery"
>
  <FieldLabel>
    Why did the delivery fail?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="failed_delivery"
  />
</Connect(Component)>
`;

exports[`renders form for incomplete_address problem_type 1`] = `
<Connect(Component)
  problem_type="incomplete_address"
>
  <FieldLabel>
    What is wrong with the addess?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="incomplete_address"
  />
</Connect(Component)>
`;

exports[`renders form for incorrect_listing problem_type 1`] = `
<Connect(Component)
  problem_type="incorrect_listing"
>
  <FieldLabel>
    Which Item?
  </FieldLabel>
  <OrderItemField
    name="order_item"
  />
  <FieldLabel>
    What is wrong with the listed product?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="incorrect_listing"
  />
</Connect(Component)>
`;

exports[`renders form for incorrect_pricing problem_type 1`] = `
<Connect(Component)
  problem_type="incorrect_pricing"
>
  <FieldLabel>
    Which Item?
  </FieldLabel>
  <OrderItemField
    name="order_item"
  />
  <FieldLabel>
    What should the price be?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="correct_price"
  />
</Connect(Component)>
`;

exports[`renders form for other problem_type 1`] = `
<Connect(Component)
  problem_type="other"
>
  <FieldLabel>
    What is the issue?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="other"
  />
</Connect(Component)>
`;

exports[`renders form for out_of_stock problem_type 1`] = `
<Connect(Component)
  problem_type="out_of_stock"
>
  <FieldLabel>
    Which Item?
  </FieldLabel>
  <OrderItemField
    name="order_item"
  />
  <FieldLabel>
    What is a good replacement item?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="replacement_item"
  />
  <FieldLabel>
    What is the price of the replacement item?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="replacement_price"
  />
</Connect(Component)>
`;

exports[`renders form for suspected_fraud problem_type 1`] = `
<Connect(Component)
  problem_type="suspected_fraud"
>
  <FieldLabel>
    Why do you believe this is a fraudulent order?
  </FieldLabel>
  <FreeResponseField
    is_required={true}
    name="suspected_fraud"
  />
</Connect(Component)>
`;
